src/rdf4cpp/rdf/IRIFactory.cpp

```diff
@@ -0,0 +1,360 @@
+    auto i = prefixes.find(prefix);
+    if (i == prefixes.end())
+        return nonstd::make_unexpected(IRIFactoryError::UnknownPrefix);
+    std::string deref {i->second};
+    deref.append(local);
+    if (IRIView{deref}.is_relative())
+        return from_relative(deref, storage);
+    return create_and_validate(deref, storage);
+}
+
+nonstd::expected<IRI, IRIFactoryError> IRIFactory::create_and_validate(std::string_view iri, storage::node::NodeStorage &storage) noexcept {
+    auto e = IRIView{iri}.quick_validate();
+    if (e != IRIFactoryError::Ok)
+        return nonstd::make_unexpected(e);
+    return IRI{iri, storage};
+}
+
+void IRIFactory::assign_prefix(std::string_view prefix, std::string_view expanded) {
+    std::string pre{prefix};
+    prefixes[pre] = expanded;
+}
+void IRIFactory::clear_prefix(std::string_view prefix) {
+    prefixes.erase(prefix);
+}
+
+std::string IRIFactory::remove_dot_segments(std::string_view path) noexcept {
+    std::string r{};
+    std::string in{path};
+    while (!in.empty()) {
+        if (in.starts_with("./")) {
+            in.erase(0, 2);
+            continue;
+        }
+        if (in.starts_with("../")) {
+            in.erase(0, 3);
+            continue;
+        }
+
+        if (in.starts_with("/./")) {
+            in.erase(0, 2);
+            continue;
+        }
+        auto seg = first_path_segment(in);
+        if (seg == "/.") {
+            in.erase(1, 1);
+            continue;
+        }
+
+        if (in.starts_with("/../")) {
+            in.erase(0, 3);
+            remove_last_path_segment(r);
+            continue;
+        }
+        if (seg == "/..") {
+            in.erase(1, 2);
+            remove_last_path_segment(r);
+            continue;
+        }
+
+        if (in == "." || in == "..") {
+            break;
+        }
+
+        r.append(seg);
+        in.erase(0, seg.length());
+    }
+
+    return r;
+}
+
+std::string_view IRIFactory::first_path_segment(std::string_view path) noexcept {
+    size_t off = 0;
+    if (path.starts_with('/'))
+        off = 1;
+    auto e = path.find('/', off);
+    return path.substr(0, e);
+}
+
+void IRIFactory::remove_last_path_segment(std::string &path) noexcept {
+    auto e = path.find_last_of('/');
+    if (e == std::string::npos)
+        return;
+    path.resize(e);
+}
+
+std::string IRIFactory::merge_paths(IRIView base, std::string_view path) noexcept {
+    if (!base.is_relative() && base.path().empty()) {
+        std::string r = "/";
+        r.append(path);
+        return r;
+    }
+    std::string r {base.path()};
+    remove_last_path_segment(r);
+    r.append(1, '/');
+    r.append(path);
+    return r;
+}
+
+std::string IRIFactory::construct(std::string_view scheme, std::optional<std::string_view> auth, std::string_view path,
+                                  std::optional<std::string_view> query, std::optional<std::string_view> frag) noexcept {
+    std::stringstream str{};
```

Comment:
Use `std::string` directly (with reserve) this is called often so it might matter for perf